### PR TITLE
Fix flaky test: authorization cancellation

### PR DESCRIPTION
**Pull request by ricardopereira**

### CODE REVIEW

File: Source/ARTAuth.m [hunk 1]
```diff
@@ -548,30 +548,43 @@ - (void)authorize:(ARTTokenParams *)tokenParams options:(ARTAuthOptions *)authOp
...
+    __block BOOL hasBeenExplicitlyCanceled = false;
     // Request always a new token
     [self.logger verbose:@"RS:%p ARTAuthInternal [authorize.%@, delegate=%@]: requesting new token", _rest, authorizeId, lastDelegate ? @"YES" : @"NO"];
     NSObject<ARTCancellable> *task;
     self->_authorizationsCount += 1;
     task = [self _requestToken:currentTokenParams withOptions:replacedOptions callback:^(ARTTokenDetails *tokenDetails, NSError *error) {
         self->_authorizationsCount -= 1;
 
-        void (^successBlock)(void) = ^{
+        void (^successCallbackBlock)(void) = ^{
```
Comment (QuintinWillison) [inline Source/ARTAuth.m:559]: The block pointers should be `const` (they're never explicitly set to `nil` anywhere, as they just fall out of scope, so we should make that clear).

File: Source/ARTAuth.m [hunk 2]
```diff
@@ -548,30 +548,43 @@ - (void)authorize:(ARTTokenParams *)tokenParams options:(ARTAuthOptions *)authOp
     __weak id<ARTAuthDelegate> lastDelegate = self.delegate;
 
     NSString *authorizeId = [[NSUUID new] UUIDString];
+    __block BOOL hasBeenExplicitlyCanceled = false;
```
Comment (QuintinWillison) [inline Source/ARTAuth.m:551]:
```suggestion
    __block BOOL hasBeenExplicitlyCanceled = NO;
```

File: Source/ARTAuth.m [hunk 3]
```diff
@@ -581,36 +594,39 @@ - (void)authorize:(ARTTokenParams *)tokenParams options:(ARTAuthOptions *)authOp
...
             }];
         }
         else {
-            successBlock();
+            successCallbackBlock();
         }
     }];
 
     [_cancelationEventEmitter once:^(ARTErrorInfo * _Nullable error) {
+        hasBeenExplicitlyCanceled = true;
```
Comment (QuintinWillison) [inline Source/ARTAuth.m:629]:
```suggestion
        hasBeenExplicitlyCanceled = YES;
```

File: Spec/Auth.swift [hunk 4]
```diff
@@ -3725,47 +3725,37 @@ class Auth : QuickSpec {
...
-                    expect(tokenDetailsFirst?.token).toNot(equal(tokenDetailsLast?.token))
                     expect(realtime.auth.tokenDetails).to(beIdenticalTo(tokenDetailsLast))
                     expect(realtime.auth.tokenDetails?.token).to(equal(tokenDetailsLast?.token))
 
                     if let transport = realtime.internal.transport as? TestProxyTransport, let query = transport.lastUrl?.query {
                         expect(query).to(haveParam("accessToken", withValue: realtime.auth.tokenDetails?.token ?? ""))
                     }
                     else {
-                        XCTFail("MockTransport is not working")
+                        fail("MockTransport is not working")
```
Comment (QuintinWillison) [inline Spec/Auth.swift:3758]: I'm interested as to why `XCTFail` was not suitable here. 🤔

File: Source/ARTAuth.m [hunk 5]
```diff
@@ -548,30 +548,43 @@ - (void)authorize:(ARTTokenParams *)tokenParams options:(ARTAuthOptions *)authOp
...
-        void (^successBlock)(void) = ^{
+        void (^successCallbackBlock)(void) = ^{
             [self.logger verbose:@"RS:%p ARTAuthInternal [authorize.%@]: success callback: %@", self->_rest, authorizeId, tokenDetails];
             if (callback) {
                 callback(tokenDetails, nil);
             }
         };
 
-        void (^failureBlock)(NSError *) = ^(NSError *error) {
+        void (^failureCallbackBlock)(NSError *) = ^(NSError *error) {
```
Comment (QuintinWillison) [inline Source/ARTAuth.m:566]: In this case, not only should the block pointer be `const` but the `error` arg passed to the callback should be too, for strict correctness. I've not run this by the compiler but if my rusty Objective-C memories are correct then that would like this:

```suggestion
        void (^const failureCallbackBlock)(NSError *) = ^(NSError *const error) {
```

**Review by QuintinWillison — CHANGES_REQUESTED** [review]
@lukasz-szyszkowski I'm going to assign this pull request to you so that you can push some commits to address review comments, as well as tweak anything you feel needs change yourself. Thanks.

File: Spec/Auth.swift (hunk 4 above)
Comment (lukasz-szyszkowski) [reply Spec/Auth.swift:3758]: When we're using `Quick/Nimble` framework for tests we should use methods provided by them, the `fail(...)` method is a part of `Nimble` framework. The above change is correct IMO but there is a lot of `XCTFail` usages in this class that should be changed for code cohesion.

File: Source/ARTAuth.m (hunk 1 above)
Comment (lukasz-szyszkowski) [reply Source/ARTAuth.m]: changed here https://github.com/ably/ably-cocoa/pull/1118/commits/23b90b26f1002fe137b9013d0a7b7e577e1dc8ca

File: Source/ARTAuth.m (hunk 2 above)
Comment (lukasz-szyszkowski) [reply Source/ARTAuth.m]: changed here https://github.com/ably/ably-cocoa/pull/1118/commits/23b90b26f1002fe137b9013d0a7b7e577e1dc8ca

File: Source/ARTAuth.m (hunk 3 above)
Comment (lukasz-szyszkowski) [reply Source/ARTAuth.m]: changed here https://github.com/ably/ably-cocoa/pull/1118/commits/23b90b26f1002fe137b9013d0a7b7e577e1dc8ca

File: Source/ARTAuth.m (hunk 5 above)
Comment (lukasz-szyszkowski) [reply Source/ARTAuth.m]: changed here https://github.com/ably/ably-cocoa/pull/1118/commits/23b90b26f1002fe137b9013d0a7b7e577e1dc8ca

Comment (maratal) [reply Source/ARTAuth.m]: Haven't seen that extensive usage of `const` before in objc. Basically it's mimicking swift `let` which is probably right thing to do, but definitely is not common practice. IMHO.

File: Spec/Auth.swift [hunk 6]
```diff
@@ -3725,47 +3725,37 @@ class Auth : QuickSpec {
...
                             partialDone()
                         }
+                        // One of them will be canceled by the connection:
+                        realtime.auth.authorize(callback)
+                        realtime.auth.authorize(callback)
                     }
 
-                    expect(tokenDetailsFirst?.token).toNot(equal(tokenDetailsLast?.token))
                     expect(realtime.auth.tokenDetails).to(beIdenticalTo(tokenDetailsLast))
                     expect(realtime.auth.tokenDetails?.token).to(equal(tokenDetailsLast?.token))
```
Comment (maratal) [inline Spec/Auth.swift:3753]: Originally `tokenDetailsLast` made sense only in a duet with tokenDetailsFirst, meaning "The first call was cancelled by the second, and the result contained in `realtime.auth.tokenDetails` is from the second call". Without the `tokenDetailsFirst` this check does nothing, except ensuring `realtime.auth.tokenDetails` is the same thing that was thrown in the callback param, which is not the purpose of this check as I can guess.

File: Source/ARTAuth.m (hunk 5 above)
Comment (lukasz-szyszkowski) [reply Source/ARTAuth.m]: I agree with @maratal, but it's not a PR merge blocker, and to be honest, I've never seen this before.

File: Spec/Auth.swift (hunk 6 above)
Comment (lukasz-szyszkowski) [reply Spec/Auth.swift:3740]: Ok, I've added one flag to ensure that the previous authorization callback was cancelled by next. The rest of these looks good to me.

Comment (lukasz-szyszkowski) [reply Spec/Auth.swift:3740]: changed here 7e23fe5501b4956a5f33ff070f5311a41d912aa2

Comment (maratal) [reply Spec/Auth.swift:3740]: Hmm.. that doesn't check whether the second call cancels the first one. This callback potentially can be called by the first call and set `didCancelAuthorization` to `true`. And this is what originally this test was trying to ensure library doesn't do.

Comment (lukasz-szyszkowski) [reply Spec/Auth.swift:3740]: As I understand this, in this test, one of these two requests should be cancelled, but it's no matter which one

**Review by maratal — APPROVED** [review]

**Review by lukasz-szyszkowski — APPROVED** [review]

**Review by QuintinWillison — APPROVED** [review]

**Pull request merged**